doc/srv6/srv6_vpn.md

```diff
@@ -0,0 +1,284 @@
+
+Schema:
+
+    ; holds local SID to behavior mapping
+    ; Status: Mandatory
+    
+    key = SRV6_MY_SID_TABLE:block_len:node_len:func_len:arg_len:ipv6address
+    
+    ; field = value
+    action = behavior                  ; behaviors defined for local SID
+    vrf = VRF_TABLE.key                ; VRF name for END.DT46, can be empty
+    adj = address,                     ; List of adjacencies for END.X, can be empty
+    segment = SRv6_SID_LIST.key,       ; List of segment names for END.B6.ENCAP, can be empty
+    source  = address,                 ; List of src addrs for encap for END.B6.ENCAP
+    
+
+![image](images/4MYSID.png)
+
+## Segment enhancement
+
+There is an enhancement added for handling segments. When the first sid in a segment matches some SRv6 MY\_SID and the MY\_SID's action is End.X, an attribute SAI\_SRV6\_SIDLIST\_ATTR\_NEXT\_HOP\_ID will be set. The value of this attribute is set to MY\_SID's nexthop id. The SAI\_SRV6\_SIDLIST\_ATTR\_SEGMENT\_LIST of this segment will not contain the first sid.
+
+## VPN Routes
+
+SRv6 VPN routes handling is added on the top of existing SRv6 routes handling. When a SRv6 VPN routes are created, SRv6 VPN routes contains VPN\_SID. SRv6Orch would check if it needs to create a corresponding SRv6 tunnel for the given BGP NH, and create the SRv6 tunnel if needed. Each SRv6 Tunnel points to an encap mapper. 
+
+The encap mapper is used to store all the VPN\_SIDs, which SRv6 VPN routes would use via this BGP NH. The key for this tunnel map is Prefix AGG\_ID. SRv6Orch allocates a Prefix AGG\_ID for a given BGP NH, and store Prefix AGG\_ID with VPN\_SID binding in encap mapper table. SRv6 tunnel would be used in SRv6 NH. SRv6 NH also points to created SID list if a set of selected IGP paths are specified via SRv6 policy. The detail SAI forwarding chain information for SRv6 VPN could be found in the above SAI section. 
+
+## ![image](images/5VPNRoute.png)VPN Routes Scale Enhancements
+
+SONiC's current design focuses on data center switch's use cases. The number of routes are limited on these types of devices. Some inefficient memory usage exists in current code base. SRv6 VPN would bring in VPN routes to the million level. These inefficient memory usage would not only bring up device's memory usage, but also slow down the BGP loading time for bringing up million routes. Therefore, we need to reevaluate the memory usage for route objects and reduce the overall (SONiC/FRR/SAI) memory footprint per route to below 2KB as a starting point. 
+
+We found that when sending 2M routes to orchagent, the orchagent will use about 3.5G of memory. Most of the memory is used by saimeta. the memory in saimeta is used for routing checks. In order to reduce memory usage and improving performance, we added a flag in the saimeta to decide whether to skip this routing check. By skipping this check, the memory usage will be reduced to 1.4G with 2M routes.
```

Comment:
This one really should be separated out from srv6 HLD. There definitely are some potential good improvements that can help routing scale with respect to memory and thus would be good to track that in overall routing specific scale enhancements

Reply:
I would like to keep it here to point out the limitation on current SONiC code base. It is a required enhancement for making it as a deployable solution.